### PR TITLE
make eps a parameter of optimisers

### PR DESCRIPTION
**Pull request by cossio**
Makes epsilon a parameter of each optimiser. Closes https://github.com/FluxML/Flux.jl/issues/1818.

### PR Checklist

- [X] Tests are added

### CODE REVIEW

File: src/optimise/optimisers.jl [hunk 1]
```diff
@@ -133,16 +133,17 @@ opt = RMSProp(0.002, 0.95)
 mutable struct RMSProp <: AbstractOptimiser
   eta::Float64
   rho::Float64
+  eps::Float64
   acc::IdDict
 end
 
-RMSProp(η = 0.001, ρ = 0.9) = RMSProp(η, ρ, IdDict())
+RMSProp(η = 0.001, ρ = 0.9, ϵ = ϵ) = RMSProp(η, ρ, ϵ, IdDict())
```
Comment (darsnack) [inline src/optimise/optimisers.jl:140]: This is what we use in Optimisers.jl.

```suggestion
RMSProp(η = 0.001, ρ = 0.9, ϵ = eps(typeof(η))) = RMSProp(η, ρ, ϵ, IdDict())
```

File: src/optimise/optimisers.jl [hunk 2]
```diff
@@ -166,10 +167,11 @@ opt = ADAM(0.001, (0.9, 0.8))
 mutable struct ADAM <: AbstractOptimiser
   eta::Float64
   beta::Tuple{Float64,Float64}
+  eps::Float64
   state::IdDict
 end
 
-ADAM(η = 0.001, β = (0.9, 0.999)) = ADAM(η, β, IdDict())
+ADAM(η = 0.001, β = (0.9, 0.999), ϵ = ϵ) = ADAM(η, β, ϵ, IdDict())
```
Comment (darsnack) [inline src/optimise/optimisers.jl:174]:
```suggestion
ADAM(η = 0.001, β = (0.9, 0.999), ϵ = eps(typeof(η))) = ADAM(η, β, ϵ, IdDict())
```

File: src/optimise/optimisers.jl [hunk 3]
```diff
@@ -207,10 +209,11 @@ opt = RADAM(0.001, (0.9, 0.8))
 mutable struct RADAM <: AbstractOptimiser
   eta::Float64
   beta::Tuple{Float64,Float64}
+  eps::Float64
   state::IdDict
 end
 
-RADAM(η = 0.001, β = (0.9, 0.999)) = RADAM(η, β, IdDict())
+RADAM(η = 0.001, β = (0.9, 0.999), ϵ = ϵ) = RADAM(η, β, ϵ, IdDict())
```
Comment (darsnack) [inline src/optimise/optimisers.jl:216]:
```suggestion
RADAM(η = 0.001, β = (0.9, 0.999), ϵ = eps(typeof(η))) = RADAM(η, β, ϵ, IdDict())
```

File: src/optimise/optimisers.jl [hunk 4]
```diff
@@ -256,10 +259,11 @@ opt = AdaMax(0.001, (0.9, 0.995))
 mutable struct AdaMax <: AbstractOptimiser
   eta::Float64
   beta::Tuple{Float64,Float64}
+  eps::Float64
   state::IdDict
 end
 
-AdaMax(η = 0.001, β = (0.9, 0.999)) = AdaMax(η, β, IdDict())
+AdaMax(η = 0.001, β = (0.9, 0.999), ϵ = ϵ) = AdaMax(η, β, ϵ, IdDict())
```
Comment (darsnack) [inline src/optimise/optimisers.jl:266]:
```suggestion
AdaMax(η = 0.001, β = (0.9, 0.999), ϵ = eps(typeof(η))) = AdaMax(η, β, ϵ, IdDict())
```

File: src/optimise/optimisers.jl [hunk 5]
```diff
@@ -298,10 +302,11 @@ opt = OADAM(0.001, (0.9, 0.995))
 mutable struct OADAM <: AbstractOptimiser
   eta::Float64
   beta::Tuple{Float64,Float64}
+  eps::Float64
   state::IdDict
 end
 
-OADAM(η = 0.001, β = (0.5, 0.9)) = OADAM(η, β, IdDict())
+OADAM(η = 0.001, β = (0.5, 0.9), ϵ = ϵ) = OADAM(η, β, ϵ, IdDict())
```
Comment (darsnack) [inline src/optimise/optimisers.jl:309]:
```suggestion
OADAM(η = 0.001, β = (0.5, 0.9), ϵ = eps(typeof(η))) = OADAM(η, β, ϵ, IdDict())
```

File: src/optimise/optimisers.jl [hunk 6]
```diff
@@ -340,16 +345,17 @@ opt = ADAGrad(0.001)
 """
 mutable struct ADAGrad <: AbstractOptimiser
   eta::Float64
+  eps::Float64
   acc::IdDict
 end
 
-ADAGrad(η = 0.1) = ADAGrad(η, IdDict())
+ADAGrad(η = 0.1, ϵ = ϵ) = ADAGrad(η, ϵ, IdDict())
```
Comment (darsnack) [inline src/optimise/optimisers.jl:352]:
```suggestion
ADAGrad(η = 0.1, ϵ = eps(typeof(η))) = ADAGrad(η, ϵ, IdDict())
```

File: src/optimise/optimisers.jl [hunk 7]
```diff
@@ -371,18 +377,19 @@ opt = ADADelta(0.89)
 """
 mutable struct ADADelta <: AbstractOptimiser
   rho::Float64
+  eps::Float64
   state::IdDict
 end
 
-ADADelta(ρ = 0.9) = ADADelta(ρ, IdDict())
+ADADelta(ρ = 0.9, ϵ = ϵ) = ADADelta(ρ, ϵ, IdDict())
```
Comment (darsnack) [inline src/optimise/optimisers.jl:384]:
```suggestion
ADADelta(ρ = 0.9, ϵ = eps(typeof(ρ))) = ADADelta(ρ, ϵ, IdDict())
```

File: src/optimise/optimisers.jl [hunk 8]
```diff
@@ -409,22 +416,23 @@ opt = AMSGrad(0.001, (0.89, 0.995))
 mutable struct AMSGrad <: AbstractOptimiser
   eta::Float64
   beta::Tuple{Float64, Float64}
+  eps::Float64
   state::IdDict
 end
 
-AMSGrad(η = 0.001, β = (0.9, 0.999)) = AMSGrad(η, β, IdDict())
+AMSGrad(η = 0.001, β = (0.9, 0.999), ϵ = ϵ) = AMSGrad(η, β, ϵ, IdDict())
```
Comment (darsnack) [inline src/optimise/optimisers.jl:423]:
```suggestion
AMSGrad(η = 0.001, β = (0.9, 0.999), ϵ = eps(typeof(η))) = AMSGrad(η, β, ϵ, IdDict())
```

File: src/optimise/optimisers.jl [hunk 9]
```diff
@@ -449,10 +457,11 @@ opt = NADAM(0.002, (0.89, 0.995))
 mutable struct NADAM <: AbstractOptimiser
   eta::Float64
   beta::Tuple{Float64, Float64}
+  eps::Float64
   state::IdDict
 end
 
-NADAM(η = 0.001, β = (0.9, 0.999)) = NADAM(η, β, IdDict())
+NADAM(η = 0.001, β = (0.9, 0.999), ϵ = ϵ) = NADAM(η, β, ϵ, IdDict())
```
Comment (darsnack) [inline src/optimise/optimisers.jl:464]:
```suggestion
NADAM(η = 0.001, β = (0.9, 0.999), ϵ = eps(typeof(η))) = NADAM(η, β, ϵ, IdDict())
```

File: src/optimise/optimisers.jl [hunk 10]
```diff
@@ -515,17 +524,18 @@ opt = AdaBelief(0.001, (0.9, 0.8))
 mutable struct AdaBelief
   eta::Float64
   beta::Tuple{Float64,Float64}
+  eps::Float64
   state::IdDict
 end
 
-AdaBelief(η = 0.001, β = (0.9, 0.999)) = AdaBelief(η, β, IdDict())
+AdaBelief(η = 0.001, β = (0.9, 0.999), ϵ = ϵ) = AdaBelief(η, β, ϵ, IdDict())
```
Comment (darsnack) [inline src/optimise/optimisers.jl:531]:
```suggestion
AdaBelief(η = 0.001, β = (0.9, 0.999), ϵ = eps(typeof(η))) = AdaBelief(η, β, ϵ, IdDict())
```

**Review by darsnack — CHANGES_REQUESTED** [review]
Only other comment is to call the field `epsilon` since that's the name we chose in Optimisers.jl (to avoid confusion with the function of the same name).

File: src/optimise/optimisers.jl (hunk 1 above)
Comment (cossio) [reply src/optimise/optimisers.jl]: I didn't want to change the value of `eps` Flux is currently using, so as to not change behavior.
So I continue to use to the value of `1e-8` here.
But I did use the type of eta as you suggest.

**Comment by cossio** [comment]
@darsnack  I didn't want to change the value of eps Flux is currently using, so as to not change behavior. So I prefer not to use `eps(typeof(eta))` as you suggest. I continue to use to the value of 1e-8 here.
But I did use the type of eta as you suggest.

Using `eps(typeof(eta))` is technically a breaking change so it could be adressed in a separate PR.

**Review by darsnack — APPROVED** [review]
Fair enough, I didn't realize the type was hard coded to `Float64`.

**Comment by cossio** [comment]
Merge?

**Comment by darsnack** [comment]
Failure seems related to #1808 (@ToucheSir)?

**Comment by ToucheSir** [comment]
Yes, https://github.com/FluxML/Flux.jl/issues/1804 specifically. I still haven't been able to repro the issue locally (will try replicating the Buildkite env when I have time in a couple of days). Feel free to `@test_skip` that one if you need any PRs in sooner, otherwise help would be much appreciated on figuring out the CI failure.

**Comment by cossio** [comment]
Merge this? buildkite failure looks unrelated to this PR, and all other tests pass.

**Comment by darsnack** [comment]
@CarloLucibello @DhairyaLGandhi I think an org owner needs to merge.

**Comment by DhairyaLGandhi** [comment]
This was a bit too wide a change for the api and can break a bunch of code...

**Comment by darsnack** [comment]
> can break a bunch of code

How so? It adds an optional positional argument that has the same default value as before.

**Comment by DhairyaLGandhi** [comment]
It adds a field to the optimisers, any code using the default "struct" constructors would break.

**Comment by ToucheSir** [comment]
Sorry, I should've caught that after https://github.com/FluxML/Flux.jl/pull/1778. Shall we quarantine both off for 0.13?

**Comment by DhairyaLGandhi** [comment]
#1778 is fine still because the disruption is limited to one type with a clear way of making a case for a non breaking type.

**Comment by cossio** [comment]
Sorry I didn't realize that. To be clear, what you mean is that a call like:

```
RMSProp(η, ρ, IdDict())
```

no longer works, because it now needs to be replaced by:

```
RMSProp(η, ρ, ϵ, IdDict())
```

right?

A possible solution is to make another PR (overwriting this one), which makes ϵ an optional keyword argument. Would that work?

```
RMSProp(η, ρ, IdDict(); ϵ = ϵ)
```

**Comment by DhairyaLGandhi** [comment]
Right

**Comment by ToucheSir** [comment]
Well, I'm still not sure if adding struct fields constitutes a SemVer break, as (de)serialization is affected. If anyone is in the know about this, please do leave a comment.

**Comment by darsnack** [comment]
@DhairyaLGandhi thanks for the explanation, sorry I missed that.

How frequently is the `IdDict` manually being passed in though? I'd rather have a breaking change for v0.13 than complicate the API by adding a keyword just to make it not breaking.

@ToucheSir adding fields is not breaking, just that the constructors conflict.

**Comment by DhairyaLGandhi** [comment]
It's got less to do with frequency and more about necessity and API compactness. In cases where we need literal copies of gradients (for synchronisation for example) over different models it's helpful to store states and gradients as the same type for simplicity, and making users add in an epsilon fiels (especially when not needed in the optimisation step) is unnecessary.

**Comment by darsnack** [comment]
I get that, but I'm saying if the frequency is low enough (i.e. a handful of users), then it's fair to ask that handful to add the new field into their code on a breaking release. Anyone manually passing in the `IdDict` is already passing in a bunch of parameters other than the LR. This seems reasonable to me in order to not have a global constant in the package affecting the behavior of many functions.

**Comment by cossio** [comment]
> How frequently is the `IdDict` manually being passed in though? I'd rather have a breaking change for v0.13 than complicate the API by adding a keyword just to make it not breaking.

I don't see why making epsilon a keyword would complicate the API? It would be an optional keyword that most users need not even notice.

**Comment by DhairyaLGandhi** [comment]
Passing in Dict explicitly is more than passing a bunch of arguments though, it is arguably more related to the use case of the optimisers than the value of an epsilon. It's already very rare to need to change epsilon.

**Comment by darsnack** [comment]
I meant that if you are manually passing in the dict, then you are already writing `ADAM(lr, (0.9, 0.999), IdDict(...))`. No one changes the `(0.9, 0.999)` either, but these are extra arguments the user is copy-pasting in order to pass in the `IdDict`. So having to update the code to `ADAM(lr, (0.9, 0.999), 1e-8, IdDict(...))` on a breaking release seems okay to me.

**Comment by DhairyaLGandhi** [comment]
But even a long list of arguments to the constructor maintains a rational order (going by what is likely to be used and requires changing for the majority of use cases), which epsilon in the position in the pr does not since it comes before the storage. It should be after.

**Comment by darsnack** [comment]
Sure, a PR to fix the order is okay for me.

**Comment by ToucheSir** [comment]
I would favour grouping the hparams together. Perhaps if the IdDict was the first parameter in the default constructor?

**Comment by cossio** [comment]
I'd prefer
```
RMSProp(η, ρ, IdDict(); ϵ = ϵ)
```
over
```
RMSProp(η, ρ, IdDict(), ϵ)
```
In fact, having all hyper-params as keyword arguments wouldn't be bad I think,
```
RMSProp(IdDict(); η=η, ρ=ρ, ϵ=ϵ)
```
but that's another matter.

**Comment by ToucheSir** [comment]
For `ADAM` at least, I'd argue `β` is not sufficiently more worthy of tuning than `ϵ` as to warrant being before the `IdDict` than after. Therein lies the challenge though: can we agree on which hyperparams are "privileged" and which are relegated to kwargs? If not, putting the `IdDict` first by convention prevents (most) any bikeshedding there.

**Comment by cossio** [comment]
Any decision here?

**Comment by DhairyaLGandhi** [comment]
I prefer the first proposal in https://github.com/FluxML/Flux.jl/pull/1819#issuecomment-1002764038. I'd imagine `ϵ` is common amongst many optimisers, and fitting it into different slots of optimiser API might get meddlesome, having it consistently as a kwarg makes sense. Besides, while experimenting with the hyper parameters of ADAM, I'd imagine modifying `β` would happen more frequently than `ϵ` (which strictly may not be a parameter at all).

**Comment by cossio** [comment]
I can make a PR
